test/evp_pkey_dhkem_test.c

```diff
@@ -0,0 +1,871 @@
+            goto err;
+    }
+    if (!TEST_int_eq(ERR_GET_REASON(ERR_get_error()), PROV_R_INVALID_KEY))
+        goto err;
+    ret = 1;
+err:
+    EVP_PKEY_free(recippriv);
+    EVP_PKEY_CTX_free(ctx);
+    return ret;
+}
+
+static int test_ec_badauth(int tstid)
+{
+    int ret = 0;
+    EVP_PKEY *auth = NULL;
+    unsigned char enc[256];
+    unsigned char secret[256];
+    unsigned char pub[256];
+    size_t enclen = sizeof(enc);
+    size_t secretlen = sizeof(secret);
+    int encap = ((tstid & 1) == 0);
+    const TEST_ENCAPDATA *t = &ec_encapdata[TEST_KEYTYPE_P256];
+    EVP_PKEY_CTX *ctx = rctx[TEST_KEYTYPE_P256];
+
+    TEST_note("%s %s", t->curve, encap ? "Encap" : "Decap");
+    /* Set the auth public key to the point at infinity */
+    pub[0] = 0;
+    if (!TEST_ptr(auth = new_raw_private_key(t->curve, t->rpriv, t->rprivlen,
+                                             pub, 1)))
+        goto err;
+    if (encap) {
+        if (!TEST_int_eq(EVP_PKEY_auth_encapsulate_init(ctx, auth,
+                                                        opparam), 1)
+            || !TEST_int_eq(EVP_PKEY_encapsulate(ctx, enc, &enclen,
+                                                 secret, &secretlen), 0))
+            goto err;
+    } else {
+        if (!TEST_int_eq(EVP_PKEY_auth_decapsulate_init(ctx, auth, opparam), 1)
+            || !TEST_int_eq(EVP_PKEY_decapsulate(ctx, secret, &secretlen,
+                                                 t->expected_enc,
+                                                 t->expected_enclen), 0))
+            goto err;
+    }
+    if (!TEST_int_eq(ERR_GET_REASON(ERR_get_error()), PROV_R_INVALID_KEY))
+        goto err;
+    ret = 1;
+err:
+    EVP_PKEY_free(auth);
+    return ret;
+}
+
+static int test_ec_invalid_decap_enc_buffer(void)
+{
+    const TEST_ENCAPDATA *t = &ec_encapdata[TEST_KEYTYPE_P256];
+    unsigned char enc[256];
+    unsigned char secret[256];
+    size_t secretlen = sizeof(secret);
+    EVP_PKEY_CTX *ctx = rctx[0];
+
+    memcpy(enc, t->expected_enc, t->expected_enclen);
+    enc[0] = 0xFF;
+
+    return TEST_int_eq(EVP_PKEY_decapsulate_init(ctx, opparam), 1)
+           && TEST_int_eq(EVP_PKEY_decapsulate(ctx, secret, &secretlen,
+                                               enc, t->expected_enclen), 0);
+}
+
+/* ECX specific tests */
+
+/* Perform ECX DHKEM KATs */
+static int test_ecx_dhkem_derivekey(int tstid)
+{
+    int ret;
+    OSSL_PARAM params[2];
+    EVP_PKEY_CTX *genctx;
+    EVP_PKEY *pkey = NULL;
+    unsigned char pubkey[64];
+    unsigned char privkey[64];
+    unsigned char masked_priv[64];
+    size_t pubkeylen = 0, privkeylen = 0;
+    const TEST_DERIVEKEY_DATA *t = &ecx_derivekey_data[tstid];
+
+    memcpy(masked_priv, t->priv, t->privlen);
+    if (OPENSSL_strcasecmp(t->curvename, "X25519") == 0) {
+        /*
+         * The RFC test vector seems incorrect since it is not in serialized form,
+         * So manually do the conversion here for now.
+         */
```

Comment:
Have we raised an errata entry for this against the RFC?

Reply:
Yes I have had a discussion with the authors and raised an errata. I assume other implementations must just be testing the public key (since it is derived from the private).